bootstrap/api/v1alpha1/rke2config_types.go

```diff
@@ -18,47 +18,624 @@ package v1alpha1
+	// +optional
+	EtcdSnapshotName string `json:"etcdSnapshotName,omitempty"`
+
+	// EtcdSnapshotScheduleCron Snapshot interval time in cron spec. eg. every 5 hours '* */5 * * *' (default: "0 */12 * * *")
+	// +optional
+	EtcdSnapshotScheduleCron string `json:"etcdSnapshotScheduleCron,omitempty"`
+
+	// EtcdSnapshotRetention Number of snapshots to retain Default: 5 (default: 5)
+	// +optional
+	EtcdSnapshotRetention string `json:"etcdSnapshotRetention,omitempty"`
+
+	// EtcdSnapshotDir Directory to save db snapshots. (Default location: ${data-dir}/db/snapshots)
+	// +optional
+	EtcdSnapshotDir string `json:"etcdSnapshotDir,omitempty"`
+
+	// EtcdS3 Enable backup to an S3-compatible Object Store
+	// +optional
+	EtcdS3 EtcdS3 `json:"etcdS3,omitempty"`
+}
+
+// Registry is registry settings including mirrors, TLS, and credentials
+type Registry struct {
+
+	// Mirrors are namespace to mirror mapping for all namespaces.
+	Mirrors map[string]Mirror `yaml:"mirrors" json:"mirrors,omitempty"`
+
+	// Configs are configs for each registry.
+	// The key is the FDQN or IP of the registry.
+	Configs map[string]RegistryConfig `yaml:"configs" json:"configs,omitempty"`
+}
+
+// Mirror contains the config related to the registry mirror
+type Mirror struct {
+
+	// Endpoints are endpoints for a namespace. CRI plugin will try the endpoints
+	// one by one until a working one is found. The endpoint must be a valid url
+	// with host specified.
+	// The scheme, host and path from the endpoint URL will be used.
+	Endpoints []string `yaml:"endpoint" json:"endpoint,omitempty"`
+
+	// Rewrites are repository rewrite rules for a namespace. When fetching image resources
+	// from an endpoint and a key matches the repository via regular expression matching
+	// it will be replaced with the corresponding value from the map in the resource request.
+	Rewrites map[string]string `yaml:"rewrite" json:"rewrite,omitempty"`
+}
+
+// RegistryConfig contains configuration used to communicate with the registry.
+type RegistryConfig struct {
+	// Auth si a reference to a Secret containing information to authenticate to the registry.
+	// The Secret must provite a username attribute and a password attribute
+	//+optional
+	AuthSecret SecretRef `json:"authSecret,omitempty"`
+	// TLS is a pair of CA/Cert/Key which then are used when creating the transport
+	// that communicates with the registry.
+	//+optional
+	TLS TLSConfig `yaml:"tls" json:"tls,omitempty"`
+}
+
+// AuthConfig contains the config related to authentication to a specific registry
+// This is not used in the CRDs but is used for marshalling the registries.yaml file
+type AuthConfig struct {
+	// Username is the username to login the registry.
+	Username string `yaml:"username" json:"username,omitempty"`
+	// Password is the password to login the registry.
+	Password string `yaml:"password" json:"password,omitempty"`
```

Comment:
Let's use an objectrefence structure https://pkg.go.dev/k8s.io/api/core/v1#ObjectReference for referencing a secret with username and password, and document that secret _must_ contain these keys.

Reply:
The `AuthConfig` struct is not used for the API, and in the `RegistryConfig.AuthSecret` attribute, I do use a `v1.ObjectReference`, so I think we can resolve this item ?

Reply:
+1, thanks a lot!